src/test/ui/let-else/let-else-drop-order.rs

```diff
@@ -0,0 +1,134 @@
+// run-pass
+// if and if let are currently inconsistent in terms of when they drop temporaries.
+// This test is meant to demonstrate the difference and show on which side
+// of the inconsistency let else and if let let chains fall upon.
+// The test bases on the "Side note" from this hackmd:
+// https://hackmd.io/@nikomatsakis/r1CpZqc8c#Side-note
+
+#![feature(let_else)]
+#![feature(let_chains)]
+
+use std::cell::RefCell;
+use std::fmt::Display;
+
+thread_local! {
+    pub static MSGS: RefCell<String> = RefCell::new(String::new());
+}
+
+fn add_msg(s: impl Display) {
+    MSGS.with(|d| {
+        *d.borrow_mut() += &format!("{s}\n");
+    });
+}
+
+fn get_msgs() -> String {
+    MSGS.with(|d| d.borrow().clone())
+}
+
+macro_rules! p {
+    ($($arg:tt),*$(,)?) => {
+        println!($($arg),*);
+        add_msg(format!($($arg),*));
+    };
+}
+
+#[derive(Default)]
+struct Droppy(u32);
+
+impl Drop for Droppy {
+    fn drop(&mut self) {
+        p!("    I dropped");
+    }
+}
+
+fn foo(v: u32) {
+    p!("v is {v}");
+    p!("  if:");
+    {
+        if Droppy(v).0 == 0 {
+            p!("    if body");
+        } else {
+            p!("    else body");
+        }
+    }
+
+    p!("  if let:");
+    {
+        if let 0 = Droppy(v).0 {
+            p!("    if body");
+        } else {
+            p!("    else body");
+        }
+    }
+
+    p!("  if let chain with let:");
+    {
+        if let 0 = Droppy(v).0 && true {
+            p!("    if body");
+        } else {
+            p!("    else body");
+        }
+    }
+
+    p!("  if let chain with bool:");
+    {
+        if let Some(0) = Some(0) && Droppy(v).0 == 0 {
+            p!("    if body");
+        } else {
+            p!("    else body");
+        }
+    }
+
+    p!("  let else:");
+    loop {
+        let 0 = Droppy(v).0 else {
+            p!("    else body");
+            break;
+        };
+        p!("    if body");
+        break;
+    }
+}
+
+fn main() {
+    foo(0);
+    foo(1);
+    assert_eq!(get_msgs(), EXPECTED_MSGS);
+}
+
+// If you want to update the expected msgs, the easiest way is to copy the stdout
+// and paste it here.
+static EXPECTED_MSGS: &str = "\
```

Comment:
We have check-stdout which produces a .stdout file and checks that against error patterns and such - maybe a better mechanism here?

Reply:
Originally I've [thought so too](https://github.com/rust-lang/rust/issues/98672#issuecomment-1183281050) but I tried both `check-stdout` out as well as adding an empty/wrong .stdout file in the hopes that compiletest corrects it upon bless, but it didn't. The problem is that .stdout files pertain to *compiler* output, not *runtime* output of the test.

So I looked again into the implementation of compiletest in preparation to tell you that no such mechanism exists, but apparently there *is* one, it's just creating .run.stdout files instead, and is activated by the `check-run-results` directive. So thanks for asking about it, it makes the test way simpler :).